firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/MutationSquasher.java

```diff
@@ -0,0 +1,249 @@
+import com.google.firebase.firestore.model.SnapshotVersion;
+import com.google.firestore.v1.Value;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Squashes multiple mutations applied on a document into a final mutation. It is guaranteed when
+ * the final mutation is applied on a document, it is as if the entire mutation list has been
+ * applied.
+ */
+public class MutationSquasher {
+
+  /** The key of the document whose mutations will be squashed */
+  private DocumentKey key;
+  /** Whether the document in question exists after `currentSquashed` is updated */
+  private boolean exists;
+  /** The mutation that is squashed from all the mutations it has seen. */
+  private Mutation currentSquashed = null;
+
+  public MutationSquasher(DocumentKey key, boolean exists) {
+    this.key = key;
+    this.exists = exists;
+  }
+
+  /** Returns the squashed mutation */
+  public Mutation getMutation() {
+    return currentSquashed;
+  }
+
+  /** Squashes the given mutation */
+  public void squash(Mutation mutation) {
+    if (mutation instanceof SetMutation) {
+      currentSquashed = mutation;
+      exists = true;
+    } else if (mutation instanceof DeleteMutation) {
+      currentSquashed = mutation;
+      exists = false;
+    } else if (currentSquashed == null) {
+      squashPatchMutationWithNull((PatchMutation) mutation);
+    } else if (currentSquashed instanceof DeleteMutation) {
+      squashPatchMutationWithDelete((DeleteMutation) currentSquashed, (PatchMutation) mutation);
+    } else if (currentSquashed instanceof SetMutation) {
+      squashPatchMutationWithSet((SetMutation) currentSquashed, (PatchMutation) mutation);
+      exists = true;
+    } else {
+      squashPatchMutationWithPatch((PatchMutation) currentSquashed, (PatchMutation) mutation);
+    }
+  }
+
+  private void squashPatchMutationWithNull(PatchMutation mutation) {
+    if (exists) {
+      currentSquashed = mutation;
+    } else {
+      MutableDocument doc = MutableDocument.newNoDocument(key, SnapshotVersion.NONE);
+      // Skip given mutation if it does not apply to no document.
+      if (mutation.getPrecondition().isValidFor(doc)) {
+        currentSquashed = mutation;
+        exists = true;
+      }
+    }
+  }
+
+  private void squashPatchMutationWithDelete(
+      DeleteMutation previouslySquashed, PatchMutation mutation) {
+    MutableDocument doc = MutableDocument.newNoDocument(mutation.getKey(), SnapshotVersion.NONE);
+    // Skip given mutation if it does not apply to no document.
+    if (!mutation.getPrecondition().isValidFor(doc)) {
+      return;
+    }
+
+    PatchMutation patchWithNoTransform =
+        new PatchMutation(
+            mutation.getKey(), mutation.getValue(), mutation.getMask(), mutation.getPrecondition());
+    patchWithNoTransform.applyToLocalView(doc, Timestamp.now());
+
+    // A Patch on a Delete turns into Set.
```

Comment:
Is this case not also invalid? A patch on a delete should get rejected, or am I misunderstanding?